### DISCUSSION TITLE
Unsure if RO/NC branch mode flag is being ignored, or am I misunderstanding eplus create policy

### DISCUSSION THREAD

**Discussion by Thomas-Geraghty** [trapexit/mergerfs#1251] [Q&A] [answered]
Hi there, first of all - thanks for taking the time to take a look at this. I'm not really sure if I'm just misunderstanding the documentation and some of the discussions/issues raised on here, or I'm facing a bug - but here goes!

My use case is:
 - Remote drive mounted using rclone to /mnt/remote with 'read-only' flag set in mount command
 - Multiple HDDs in JBOD, combined together using mergerfs, in combination with the data stored on the remote drive.
 - I would like to keep any new files/folders in the same path on the same drive, but if the path only exists on the readonly remote drive, it should just create a new path on one of the other branches and begin adding to that instead.

I've got a mount command like this

`mergerfs /mnt/drives/hdd\*/archive:/mnt/drives/remote/archive=RO /mnt/pools/archive -o 
func.getattr=newest,
category.create=eplus,
cache.files=partial,
noforget,
nonempty,
dropcacheonclose=true`

SoI've got an underlying file structure like this
```
/mnt/
└── drives/
    ├── hdd1/
    │   └── archive/
    └── remote/
        └── archive/
            └── repository-1/
```

If I try write a file to the mergerfs mount folder, into the 'repository-1' folder
`touch /mnt/pools/archive/repository-1/testFile`
I will get an error saying
`touch: cannot touch '/mnt/pools/archive/repository-1/testFile': Read-only file system`

So obviously `eplus` creation policy means that mergerfs should try respect and use branch that already has the existing path, which in this case would be the `remote` branch. But then at the same time, that branch is set to `=RO`, which means it should be ignored for the create action right, so why is it still trying to add the file into the readonly `remote` branch?

Could you just clarify how this works? Does the =RO / =NC branch mode block any attempt to create a file on that branch? And if it does, doesn't eplus respect this and realise it cannot write to that branch so much fallback to another drive? Is something wrong with my command maybe?

Just for reference, I am on `mergerfs v2.37.1`

**Comment by trapexit** [comment]
There really isn't more to say than what is in the docs. https://github.com/trapexit/mergerfs#filtering

For each branch:
* filter out anything with =RO or =NC on file create.
* filter out branches with less than minfreespace
* filter out branches that don't have the relative base path
* of those left... pick the one with the least free space

The relative path to `testFile` is `repository-1`. So as configured it only considered branches with that relative path. Which none of your hdds have. It's behaving exactly as expected. _"Of all the branches on which the relative path exists choose the branch with the least used space."_

`ep` policies are very niche. For people who want to manually control where files are created.

https://github.com/trapexit/mergerfs#what-policies-should-i-use

**Comment by Thomas-Geraghty** [comment]
Thanks for responding so quick @trapexit, appreciate all your efforts!

Is it possible to configure so that if the existing path is readonly, fallback and create the path/file on a different branch? I guess maybe something similar to `msplus` but rather than checking the parent directory, instead trying on a different branch?

I think I've got a similar use-case to this user here
https://github.com/trapexit/mergerfs/issues/840
edit: and here https://github.com/trapexit/mergerfs/discussions/1034#discussioncomment-3398930

where I could double mount mergerfs to achieve want I want (albeit this is not recommended)
- Instance A of mergerfs set to to `lus` policy that only merges the hdd* drives
- Instance B of mergerfs that merges instance A and the remote drive in `ff` policy

Does this maybe make my use case a bit more clear? I don't see an option offered policy wise that really covers this

**Reply by trapexit** [reply]
policies don't choose where to create directories. They choose which branch to use. Creation of directories is done as needed after the policy runs. If you want mergerfs to pick the other branch choose almost anything else but `ep` policies. You've very explicitly told it not to consider any branch that doesn't have the relative path... so that's what it is doing. Literally anything else would do what you want. A RO or NC branch will always be ignored for creates.

**Reply by trapexit** [reply]
I don't see where you are asking about nested / hierarchical / list of list behaviors. You said above you just want to ignore remote branch. It will. It is. Your problem is you are also telling it to ignore other branches. Just don't do that.

**Reply by Thomas-Geraghty** [reply]
Okay maybe I'm confusing you here

- /mnt/drives/remote - should have no files written to, no changes made etc. Purely read-only.
- /mnt/drives/hdd* - can have files written to, should use eplus policy for choosing which drive for writing files to.

In my current setup, it doesn't work because it can see the path on the `/mnt/drives/remote` but isn't allowed to write there. If I set it my create policy to `lus` it'll write new files to /mnt/drives/hdd* but it'll spread them across all the drives, which i'd rather it kept them together and instead spread new top-level directories.

**Reply by trapexit** [reply, marked as answer]
Then use a `msp*` policy.

**Reply by Thomas-Geraghty** [reply]
Fantastic, that solved it. Thank you so much for your help.
